client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,343 @@
+import { getVariant } from "../data/gameData";
+import * as deckRules from "../rules/deck";
+import { GameAction } from "../types/actions";
+import CardState from "../types/CardState";
+import GameMetadata from "../types/GameMetadata";
+import Variant from "../types/Variant";
+
+export default function cardDeductionReducer(
+  deck: readonly CardState[],
+  oldDeck: readonly CardState[],
+  action: GameAction,
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+): CardState[] {
+  switch (action.type) {
+    case "cardIdentity":
+    case "clue":
+    case "discard":
+    case "play":
+    case "draw": {
+      return makeDeductions(deck, oldDeck, hands, metadata);
+    }
+    default: {
+      return Array.from(deck);
+    }
+  }
+}
+
+function makeDeductions(
+  deck: readonly CardState[],
+  oldDeck: readonly CardState[],
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+) {
+  const newDeck: CardState[] = Array.from(deck);
+  const variant = getVariant(metadata.options.variantName);
+  const cardCountMap = getCardCountMap(variant);
+
+  // We need to calculate our own unknown cards first because those possibilities will be needed for
+  // pretending like we know what the other players see.
+  calculatePlayerPossibilities(
+    metadata.ourPlayerIndex,
+    metadata.ourPlayerIndex,
+    hands,
+    newDeck,
+    oldDeck,
+    cardCountMap,
+  );
+  for (let playerIndex = 0; playerIndex < hands.length; playerIndex++) {
+    if (playerIndex !== metadata.ourPlayerIndex) {
+      calculatePlayerPossibilities(
+        playerIndex,
+        metadata.ourPlayerIndex,
+        hands,
+        newDeck,
+        oldDeck,
+        cardCountMap,
+      );
+    }
+  }
+  return newDeck;
+}
+
+function calculatePlayerPossibilities(
+  playerIndex: number,
+  ourPlayerIndex: number,
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  oldDeck: readonly CardState[],
+  cardCountMap: readonly number[][],
+) {
+  const cardsToCalculate = getCardsToCalculate(
+    playerIndex,
+    ourPlayerIndex,
+    hands,
+    deck,
+    oldDeck,
+  );
+
+  cardsToCalculate.forEach((card) => {
+    const possibilities = generatePossibilitiesForUnknownCards(
+      card.order,
+      deck,
+      playerIndex,
+      ourPlayerIndex,
+    );
+    let { possibleCards, possibleCardsForEmpathy } = card;
+    if (playerIndex === ourPlayerIndex) {
+      possibleCards = possibleCards.filter((possibility) =>
+        possibilityValid(possibility, possibilities, 0, cardCountMap),
+      );
+      if (playerIndex === card.location) {
+        // If the card is in our own hand then we also need to update empathy
+        if (card.revealedToPlayer[playerIndex]) {
+          // If it's revealed to us then we can safely just use possibleCards and don't
+          // need to do a separate calculation run for possibleCardsForEmpathy.
+          possibleCardsForEmpathy = possibleCards;
```

Comment:
When would a card in our hand ever be revealed to us?

Reply:
It won't, that's a bad if check. I initially had it just if it's a card in our own hand then make possibleCardsForEmpathy = possibleCards. That caused replays to have bad empathy in our own hand because possibleCards is just one thing (the real value of the card). So I attempted to fix it by saying "don't do that if the card isn't revealed to us" but that just caused everything to just go down the path of calculating empathy separately.

I've updated it to your simplified logic. Possibly at some point in the future there's a way to split those. It might not be worth the complexity though, since replays are where we take the biggest performance hit and it needs to be run separately for that.